readium/lcp/src/main/java/org/readium/r2/lcp/LcpPublicationRetriever.kt

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2023 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.lcp
+
+import android.content.Context
+import java.io.File
+import org.readium.r2.lcp.license.container.createLicenseContainer
+import org.readium.r2.lcp.license.model.LicenseDocument
+import org.readium.r2.shared.extensions.tryOrLog
+import org.readium.r2.shared.util.Url
+import org.readium.r2.shared.util.downloads.DownloadManager
+import org.readium.r2.shared.util.mediatype.FormatRegistry
+import org.readium.r2.shared.util.mediatype.MediaType
+import org.readium.r2.shared.util.mediatype.MediaTypeRetriever
+
+/**
+ * Util to acquire a protected publication from standalone LCPL's bytes.
+ */
+public class LcpPublicationRetriever(
+    context: Context,
+    private val downloadManager: DownloadManager,
+    private val mediaTypeRetriever: MediaTypeRetriever
+) {
+
+    @JvmInline
+    public value class RequestId(public val value: String)
+
+    public interface Listener {
+
+        public fun onAcquisitionCompleted(
+            requestId: RequestId,
+            acquiredPublication: LcpService.AcquiredPublication
+        )
+
+        public fun onAcquisitionProgressed(
+            requestId: RequestId,
+            downloaded: Long,
+            expected: Long?
+        )
+
+        public fun onAcquisitionFailed(
+            requestId: RequestId,
+            error: LcpException
+        )
+
+        public fun onAcquisitionCancelled(
+            requestId: RequestId
+        )
+    }
+
+    private inner class DownloadListener : DownloadManager.Listener {
+
+        override fun onDownloadCompleted(
+            requestId: DownloadManager.RequestId,
+            file: File
+        ) {
+            val lcpRequestId = RequestId(requestId.value)
+            val listenersForId = listeners[lcpRequestId].orEmpty()
+
+            val license = downloadsRepository.retrieveLicense(requestId.value)
+                ?.let { LicenseDocument(it) }
+                ?: run {
+                    listenersForId.forEach {
+                        it.onAcquisitionFailed(
+                            lcpRequestId,
+                            LcpException.wrap(
+                                Exception("Couldn't retrieve license from local storage.")
+                            )
+                        )
+                    }
+                    return
+                }
+            downloadsRepository.removeDownload(requestId.value)
+
+            val link = license.link(LicenseDocument.Rel.Publication)!!
```

Comment:
Since we know that there will always be a publication link in the contract, how about adding this in `LicenseDocument` to avoid the force unwrapping in the callers?

```kotlin
val LicenseDocument.publicationLink: Link get() = link(Rel.Publication)!!
```